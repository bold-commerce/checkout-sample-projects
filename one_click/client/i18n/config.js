import i18n from 'i18next';
import enTranslation from './en/translation.json';
import { initReactI18next } from 'react-i18next';

i18n
  .use(initReactI18next)
  .init({
    resources: {
      en: {
        translation: enTranslation
      }
    },
    lng: 'en',
    fallbackLng: 'en',
    interpolation: {
      escapeValue: false, 
    }
  });

export default i18n;