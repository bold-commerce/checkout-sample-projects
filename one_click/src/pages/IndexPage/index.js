export { default as IndexPage } from "./IndexPage";
export { default as IndexPageGuest } from "./IndexPageGuest";