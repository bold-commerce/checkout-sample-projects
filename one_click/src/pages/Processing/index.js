export { default as ProcessingOrder } from './Processing'