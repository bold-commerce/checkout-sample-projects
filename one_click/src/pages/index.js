export * from './Confirmation';
export * from './IndexPage';
export * from './Processing';
export * from './Inventory';