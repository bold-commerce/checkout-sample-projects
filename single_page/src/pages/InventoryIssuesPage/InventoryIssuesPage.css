.FieldSet.InventoryIssues__Section {
  padding: 16px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  margin: 16px 0;
}

.Checkout__InventoryIssues {
  width: 100%;
  padding: 35px;
  margin: 0 auto;
  max-width: 800px;
}

.Checkout__ConfirmButton {
  background: #4e4e4e;
  color: #fff;
}

.InventoryIssues__Section p {
  color: rgba(0,0,0,0.65);
}

.LineItem__QuantityPriceWrapper svg {
  height: 100%;
}