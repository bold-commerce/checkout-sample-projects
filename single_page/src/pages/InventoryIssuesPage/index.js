export { default as InventoryIssuesPage } from './InventoryIssuesPage';
