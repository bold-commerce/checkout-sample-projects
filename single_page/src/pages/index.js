export * from './ConfirmationPage';
export * from './IndexPage';
export * from './ProcessingPage';
