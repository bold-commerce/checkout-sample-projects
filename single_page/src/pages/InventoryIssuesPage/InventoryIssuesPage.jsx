import React, { useCallback, useMemo, useState } from 'react';
import { useLocation, useNavigate } from 'react-router';
import { useLineItems } from '@boldcommerce/checkout-react-components';
import { CheckoutSection, LineItem } from '../../components';
import { useErrorLogging } from '../../hooks'; 
import './InventoryIssuesPage.css';
import { Button } from '@boldcommerce/stacks-ui';
import { useTranslation } from 'react-i18next';

const InventoryIssuesPage = () => {
  const location = useLocation();
  const navigate = useNavigate();
  const logError = useErrorLogging();
  const { data: lineItems, updateLineItemQuantity, removeLineItem } = useLineItems();
  const inventoryIssues = location.state;
  const [loading, setLoading] = useState(false);
  const { t } = useTranslation();

  // Only run this when the component mounts since we don't want the list of items to change as it adjusts quantities
  const updatedLineItems = useMemo(() => {
    let inventoryMap = inventoryIssues.reduce((acc,curr)=> (acc[curr.platform_id]=curr.inventory_quantity,acc),{});
    const adjustedLineItems = lineItems.map((lineItem) => {
      const variantInventory = inventoryMap[lineItem.product_data.variant_id];
      let quantity = lineItem.product_data.quantity;

      if (quantity > variantInventory) {
        quantity = variantInventory;
        inventoryMap[lineItem.product_data.variant_id] = 0;
      } else {
        inventoryMap[lineItem.product_data.variant_id] -= quantity;
      }

      return {
        ...lineItem,
        product_data: {
          ...lineItem.product_data,
          quantity,
          originalQuantity: lineItem.product_data.quantity,
        },
      }
    });

    return adjustedLineItems;
  }, []);

  const lineItemList = updatedLineItems.map((lineItem) => (
    <LineItem
      title={lineItem.product_data.title}
      image={lineItem.product_data.image_url}
      quantity={lineItem.product_data.quantity}
      originalQuantity={lineItem.product_data.originalQuantity}
      price={lineItem.product_data.price}
      totalPrice={lineItem.product_data.total_price}
      lineItemKey={lineItem.product_data.line_item_key}
      readOnly
      key={lineItem.product_data.line_item_key}
    />
  ));

  const handleChanges = useCallback(async () => {
    setLoading(true);
    try {
      let results = [];
  
      for (let i = 0; i < updatedLineItems.length; i++) {
        const item = updatedLineItems[i];
        if (item.product_data.quantity === 0) {
          results.push(removeLineItem(item.product_data.line_item_key));
        } else if (item.product_data.quantity !== item.product_data.originalQuantity) {
          results.push(updateLineItemQuantity(item.product_data.line_item_key, item.product_data.quantity));
        }
      }
  
      await Promise.all(results);
      setLoading(false);
      navigate('/');
    } catch(e) {
      setLoading(false);
      logError('inventory_issues', e);
    }
  }, [updatedLineItems]);

  return (
    <div className="Checkout__InventoryIssues" role="main">
      <CheckoutSection
        className="InventoryIssues__Section"
        title={t('inventory.issues')}
      >
        <p>{t('inventory.issues_description')}</p>
      </CheckoutSection>
      <div className="InventoryIssues__List">
        {lineItemList}
      </div>
      <Button
        type="button"
        className="Checkout__ConfirmButton"
        onClick={handleChanges}
        loading={loading}
        disabled={loading}
      >
        {t('inventory.continue')}
      </Button>
    </div>
  );
};

export default InventoryIssuesPage;
