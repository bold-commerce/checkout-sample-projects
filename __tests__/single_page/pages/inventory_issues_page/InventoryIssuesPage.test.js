import React from 'react';
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { InventoryIssuesPage } from "../../../../single_page/src/pages/InventoryIssuesPage";
import { exampleUseLineItems as MOCKlineItems } from "../../../utils/hookHelpers";
import '../../../../single_page/src/i18n/config';

jest.mock('@boldcommerce/checkout-react-components', () => ({
  ...jest.requireActual('@boldcommerce/checkout-react-components'),
  useLineItems: () => MOCKlineItems,
})).mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn().mockImplementation(() => {
    return { state: [] }
  }),
  useNavigate: jest.fn().mockImplementation(() => {
    return { pathname: {}}
  })
}));

describe('InventoryIssuesPage', () => {
  test('renders InventoryIssuesPage page', () => {
    const { asFragment } = render(
      <InventoryIssuesPage/>,
      {wrapper: MemoryRouter}
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
