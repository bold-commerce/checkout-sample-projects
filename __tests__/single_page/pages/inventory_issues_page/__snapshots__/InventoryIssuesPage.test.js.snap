// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryIssuesPage renders InventoryIssuesPage page 1`] = `
<DocumentFragment>
  <div
    class="Checkout__InventoryIssues"
    role="main"
  >
    <section
      class="FieldSet InventoryIssues__Section"
    >
      <div
        class="FieldSet__Header"
      >
        <h3
          class="FieldSet__Heading"
        >
          Inventory issues
        </h3>
      </div>
      <div
        class="FieldSet__Content"
      >
        <p>
          Some Products became unavailable and your cart has been updated, We're sorry for the inconvenience.
        </p>
      </div>
    </section>
    <div
      class="InventoryIssues__List"
    >
      <div
        class="LineItem"
      >
        <div
          class="LineItem__ImageDescriptionWrapper"
        >
          <div
            class="LineItem__ProductImageWrapper"
          >
            <div
              class="stx-image"
            >
              <div
                class="stx-image--content"
              >
                <div
                  class="stx-image--holder"
                >
                  <img
                    alt="Size Large / Colour Blue"
                    class="stx-image--image"
                    src=""
                    title="Size Large / Colour Blue"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="LineItem__ProductDetails"
          >
            <div
              class="stx-details stx-details--left"
            >
              <div
                class="stx-details__title"
              >
                Size Large / Colour Blue
              </div>
              <div
                class="stx-details__description"
              />
            </div>
          </div>
        </div>
        <div
          class="LineItem__QuantityPriceWrapper"
        >
          <div
            class="LineItem__QuantityAdjustmentWrapper"
          >
            <div
              class="LineItem__LineItemQuantityWrapper"
            >
              <div
                class="LineItemQuantity__Decrement"
              />
              <div
                class="LineItemQuantity LineItemQuantity--read-only"
              >
                <span
                  class="LineItemQuantity__ReadOnlyValue"
                >
                  1
                </span>
              </div>
              <div
                class="LineItemQuantity__Increment"
              />
            </div>
            <svg
              fill="none"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.7239 7.99998L8.19526 2.47138C7.93491 2.21103 7.93491 1.78892 8.19526 1.52858C8.45561 1.26823 8.87772 1.26823 9.13807 1.52858L15.1381 7.52858C15.3984 7.78893 15.3984 8.21103 15.1381 8.47138L9.13807 14.4714C8.87772 14.7317 8.45561 14.7317 8.19526 14.4714C7.93491 14.211 7.93491 13.7889 8.19526 13.5286L13.7239 7.99998Z"
                fill="black"
                fill-opacity="0.65"
              />
              <path
                d="M13.3332 7.33331C13.7014 7.33331 13.9998 7.63179 13.9998 7.99998C13.9998 8.36817 13.7014 8.66665 13.3332 8.66665H1.33317C0.964981 8.66665 0.666504 8.36817 0.666504 7.99998C0.666504 7.63179 0.964981 7.33331 1.33317 7.33331L13.3332 7.33331Z"
                fill="black"
                fill-opacity="0.65"
              />
            </svg>
            <div
              class="LineItem__LineItemQuantityWrapper"
            >
              <div
                class="LineItemQuantity__Decrement"
              />
              <div
                class="LineItemQuantity LineItemQuantity--read-only"
              >
                <span
                  class="LineItemQuantity__ReadOnlyValue"
                >
                  1
                </span>
              </div>
              <div
                class="LineItemQuantity__Increment"
              />
            </div>
          </div>
          <div
            class="LineItem__ProductPrice"
          >
            <div
              class="stx-price stx-price--alight-center"
            >
              <p
                aria-label="Product price"
                class="stx-price__amount"
              >
                $14.00 CAD
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="LineItem"
      >
        <div
          class="LineItem__ImageDescriptionWrapper"
        >
          <div
            class="LineItem__ProductImageWrapper"
          >
            <div
              class="stx-image"
            >
              <div
                class="stx-image--content"
              >
                <div
                  class="stx-image--holder"
                >
                  <img
                    alt="Number Two"
                    class="stx-image--image"
                    src=""
                    title="Number Two"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="LineItem__ProductDetails"
          >
            <div
              class="stx-details stx-details--left"
            >
              <div
                class="stx-details__title"
              >
                Number Two
              </div>
              <div
                class="stx-details__description"
              />
            </div>
          </div>
        </div>
        <div
          class="LineItem__QuantityPriceWrapper"
        >
          <div
            class="LineItem__QuantityAdjustmentWrapper"
          >
            <div
              class="LineItem__LineItemQuantityWrapper"
            >
              <div
                class="LineItemQuantity__Decrement"
              />
              <div
                class="LineItemQuantity LineItemQuantity--read-only"
              >
                <span
                  class="LineItemQuantity__ReadOnlyValue"
                >
                  2
                </span>
              </div>
              <div
                class="LineItemQuantity__Increment"
              />
            </div>
            <svg
              fill="none"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.7239 7.99998L8.19526 2.47138C7.93491 2.21103 7.93491 1.78892 8.19526 1.52858C8.45561 1.26823 8.87772 1.26823 9.13807 1.52858L15.1381 7.52858C15.3984 7.78893 15.3984 8.21103 15.1381 8.47138L9.13807 14.4714C8.87772 14.7317 8.45561 14.7317 8.19526 14.4714C7.93491 14.211 7.93491 13.7889 8.19526 13.5286L13.7239 7.99998Z"
                fill="black"
                fill-opacity="0.65"
              />
              <path
                d="M13.3332 7.33331C13.7014 7.33331 13.9998 7.63179 13.9998 7.99998C13.9998 8.36817 13.7014 8.66665 13.3332 8.66665H1.33317C0.964981 8.66665 0.666504 8.36817 0.666504 7.99998C0.666504 7.63179 0.964981 7.33331 1.33317 7.33331L13.3332 7.33331Z"
                fill="black"
                fill-opacity="0.65"
              />
            </svg>
            <div
              class="LineItem__LineItemQuantityWrapper"
            >
              <div
                class="LineItemQuantity__Decrement"
              />
              <div
                class="LineItemQuantity LineItemQuantity--read-only"
              >
                <span
                  class="LineItemQuantity__ReadOnlyValue"
                >
                  2
                </span>
              </div>
              <div
                class="LineItemQuantity__Increment"
              />
            </div>
          </div>
          <div
            class="LineItem__ProductPrice"
          >
            <div
              class="stx-price stx-price--alight-center"
            >
              <p
                aria-label="Product price"
                class="stx-price__amount"
              >
                $32.00 CAD
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="LineItem"
      >
        <div
          class="LineItem__ImageDescriptionWrapper"
        >
          <div
            class="LineItem__ProductImageWrapper"
          >
            <div
              class="stx-image"
            >
              <div
                class="stx-image--content"
              >
                <div
                  class="stx-image--holder"
                >
                  <img
                    alt="Default Title"
                    class="stx-image--image"
                    src=""
                    title="Default Title"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="LineItem__ProductDetails"
          >
            <div
              class="stx-details stx-details--left"
            >
              <div
                class="stx-details__title"
              >
                Default Title
              </div>
              <div
                class="stx-details__description"
              />
            </div>
          </div>
        </div>
        <div
          class="LineItem__QuantityPriceWrapper"
        >
          <div
            class="LineItem__QuantityAdjustmentWrapper"
          >
            <div
              class="LineItem__LineItemQuantityWrapper"
            >
              <div
                class="LineItemQuantity__Decrement"
              />
              <div
                class="LineItemQuantity LineItemQuantity--read-only"
              >
                <span
                  class="LineItemQuantity__ReadOnlyValue"
                >
                  3
                </span>
              </div>
              <div
                class="LineItemQuantity__Increment"
              />
            </div>
            <svg
              fill="none"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.7239 7.99998L8.19526 2.47138C7.93491 2.21103 7.93491 1.78892 8.19526 1.52858C8.45561 1.26823 8.87772 1.26823 9.13807 1.52858L15.1381 7.52858C15.3984 7.78893 15.3984 8.21103 15.1381 8.47138L9.13807 14.4714C8.87772 14.7317 8.45561 14.7317 8.19526 14.4714C7.93491 14.211 7.93491 13.7889 8.19526 13.5286L13.7239 7.99998Z"
                fill="black"
                fill-opacity="0.65"
              />
              <path
                d="M13.3332 7.33331C13.7014 7.33331 13.9998 7.63179 13.9998 7.99998C13.9998 8.36817 13.7014 8.66665 13.3332 8.66665H1.33317C0.964981 8.66665 0.666504 8.36817 0.666504 7.99998C0.666504 7.63179 0.964981 7.33331 1.33317 7.33331L13.3332 7.33331Z"
                fill="black"
                fill-opacity="0.65"
              />
            </svg>
            <div
              class="LineItem__LineItemQuantityWrapper"
            >
              <div
                class="LineItemQuantity__Decrement"
              />
              <div
                class="LineItemQuantity LineItemQuantity--read-only"
              >
                <span
                  class="LineItemQuantity__ReadOnlyValue"
                >
                  3
                </span>
              </div>
              <div
                class="LineItemQuantity__Increment"
              />
            </div>
          </div>
          <div
            class="LineItem__ProductPrice"
          >
            <div
              class="stx-price stx-price--alight-center"
            >
              <p
                aria-label="Product price"
                class="stx-price__amount"
              >
                $123.00 CAD
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      aria-busy="false"
      aria-disabled="false"
      class="stx-button Checkout__ConfirmButton"
      type="button"
    >
      <span
        class="stx-button__text"
      >
        Continue with changes
      </span>
      <span
        class="stx-loading-spinner stx-button__spinner"
      />
    </button>
  </div>
</DocumentFragment>
`;
